packages/tokenbank_v1/contracts/impl/MetaTxModule.sol

```diff
@@ -0,0 +1,255 @@
+        uint256 value,
+        bytes   memory data,
+        uint256 nonce,
+        uint256 gasPrice,
+        uint256 gasLimit,
+        address gasToken,
+        bytes   memory extraHash
+        )
+        internal
+        pure
+        returns (bytes32)
+    {
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                byte(0x19),
+                byte(0),
+                from,
+                to,
+                value,
+                keccak256(data),
+                nonce,
+                gasPrice,
+                gasLimit,
+                gasToken,
+                extraHash
+                )
+            );
+
+        return keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hash));
+    }
+
+    /// @dev   Recovers the signer at a given index from a list of concatenated signatures.
+    /// @param metaTxHash The signed hash
+    /// @param signatures The concatenated signatures.
+    /// @param index The index of the signature to recover.
+    function recoverSigner(
+        bytes32      metaTxHash,
+        bytes memory signatures,
+        uint         index
+        )
+        internal
+        pure
+        returns (address)
+    {
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+        // we jump 32 (0x20) as the first slot of bytes contains the length
+        // we jump 65 (0x41) per signature
+        // for v we load 32 bytes ending with v (the first 31 come from s) then apply a mask
+        assembly {
+            r := mload(add(signatures, add(0x20, mul(0x41, index))))
+            s := mload(add(signatures, add(0x40, mul(0x41, index))))
+            v := and(mload(add(signatures, add(0x41, mul(0x41, index)))), 0xff)
+        }
+        require(v == 27 || v == 28, "");
+        return ecrecover(metaTxHash, v, r, s);
+    }
+
+    function extractMethod(bytes memory data)
+        internal
+        pure
+        returns (bytes4 method)
+    {
+        require(data.length >= 4, "INVALID_DATA");
+        assembly {
+            // data layout: {length:32}{sig:4}{_wallet:32}{...}
+            method := mload(add(data, 32))
+        }
+    }
+
+    /// @dev Save the meta-transaction to history.
+    ///      This method must throw if the transaction is not unique or the nonce is invalid.
+    /// @param wallet The target wallet.
+    /// @param nonce The nonce
+    /// @param metaTxHash The signed hash of the transaction
+    function saveExecutedMetaTx(
+        address wallet,
+        uint    nonce,
+        bytes32 metaTxHash
+        )
+        private
+    {
+        if (nonce == 0) {
+            require(!wallets[wallet].execuitedMetaTxHash[metaTxHash], "DUPLICIATE_SIGN_HASH");
+            wallets[wallet].execuitedMetaTxHash[metaTxHash] = true;
+        } else {
+            require(nonce <= wallets[wallet].nonce, "NONCE_TOO_SMALL");
+            uint nonceBlock = (nonce & 0xffffffffffffffffffffffffffffffff00000000000000000000000000000000) >> 128;
```

Comment:
I don't think you need to do the AND if you shift to the right anyway, but I guess it can't hurt. :)